src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/RuntimeTypeMetadataUpdateHandler.cs

```diff
@@ -10,16 +12,60 @@ namespace System.Reflection.Metadata
     /// <summary>Metadata update handler used to clear a Type's reflection cache in response to a metadata update notification.</summary>
     internal static class RuntimeTypeMetadataUpdateHandler
     {
-        public static void BeforeUpdate(Type? type)
+        /// <summary>Clear type caches in response to an update notification.</summary>
+        /// <param name="types">The specific types to be cleared, or null to clear everything.</param>
+        public static void BeforeUpdate(Type[]? types)
         {
-            if (type is RuntimeType rt)
+            if (RequiresClearingAllTypes(types))
             {
-                rt.ClearCache();
+                // Clear the caches on all loaded types.
+                foreach (AssemblyLoadContext alc in AssemblyLoadContext.All)
+                {
+                    foreach (Assembly assembly in alc.Assemblies)
+                    {
+                        foreach (Type type in assembly.GetTypes())
+                        {
+                            ClearCache(type);
+                        }
+                    }
+                }
+            }
+            else
+            {
+                // Clear the caches on just the specified types.
+                foreach (Type type in types)
+                {
+                    ClearCache(type);
+                }
+            }
+        }
+
+        /// <summary>Clears the cache on the specified type.</summary>
+        private static void ClearCache(Type type) => (type as RuntimeType)?.ClearCache();
+
+        /// <summary>Determines whether we need to clear the caches on all types or just the ones specified.</summary>
+        /// <param name="types">The types to examine.</param>
+        /// <returns>true if we need to clear all types; false if we can clear just the ones specified.</returns>
+        private static bool RequiresClearingAllTypes([NotNullWhen(false)] Type[]? types)
+        {
+            // If we were handed null, assume we need to clear everything.
+            if (types is null)
```

Comment:
@pranavkm, this should enable us to get clearing working from the SDK even before we know the types.  We can pass null for now and update it later when we have more specific information.